hail/python/hail/expr/functions.py

```diff
@@ -3736,20 +3755,23 @@ def deep_flatten(l):
         return result
 
     if isinstance(collection, np.ndarray):
-        nd = collection.astype(np.float64)
-        data = list(nd.flat)
-        shape = nd.shape
-    elif isinstance(collection, list):
+        if row_major is None:
+            return to_expr(collection)
+        else:
```

Comment:
It's a bit weird to me that you've essentially repeated the code in `to_expr` below. Not sure there is a better way to structure this

Reply:
Ya, the issue is that `to_expr` assumes everything it needs to coerce data to expressions is present in the data itself. There also isn't a good way to coerce python lists to NDArrays without first going through hail arrays which I don't want to do, leading to this awkward half and half. Maybe it would be best for now to leave `to_expr` alone and do everything explicitly in `_ndarray`.

Reply:
I guess another possibility is to have a `NDArrayReorder` node that allows you to change the order you impose on an `NDArray` and a simplify rule to simplify `(NDArrayReorder (MakeNDArray))`

Reply:
That also gets quite low-level, in manipulating the memory layout, and would be expensive if ever actually used in another setting. Not to mention it's quite confusing and subtle how that differs from a transpose

Reply:
I can definitely extract the repeated code, but wanted to have a more general discussion around the above.

Reply:
A reorder node might actually be quite useful down the road if we can inspect an IR and make a smart decision a priori about what the optimal memory layout for the computation would be.

Reply:
Ultimately backed out of `to_expr` for now and kept all logic in `_ndarray`. I don't think any of the above are really in the scope of this PR.